pkg/auth/providers/openldap/openldap_client.go

```diff
@@ -0,0 +1,447 @@
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/sirupsen/logrus"
+	ldapv2 "gopkg.in/ldap.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (p *openldapProvider) loginUser(credential *v3public.BasicLogin, config *v3.OpenLDAPConfig, caPool *x509.CertPool) (v3.Principal, []v3.Principal, map[string]string, error) {
+	logrus.Debug("Now generating Ldap token")
+
+	username := credential.Username
+	password := credential.Password
+
+	if password == "" {
+		return v3.Principal{}, nil, nil, httperror.NewAPIError(httperror.MissingRequired, "password not provided")
+	}
+
+	lConn, err := ldap.NewLDAPConnForOpenLDAP(config, caPool)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+	defer lConn.Close()
+
+	if !config.Enabled { // TODO testing for enabled here might not be correct. Might be better to pass in an explicit testSvcAccount bool
+		logrus.Debug("Bind service account username password")
+		if config.ServiceAccountPassword == "" {
+			return v3.Principal{}, nil, nil, httperror.NewAPIError(httperror.MissingRequired, "service account password not provided")
+		}
+		sausername := ldap.GetUserExternalID(config.ServiceAccountUsername, "")
+		err = lConn.Bind(sausername, config.ServiceAccountPassword)
+		if err != nil {
+			if ldapv2.IsErrorWithCode(err, ldapv2.LDAPResultInvalidCredentials) {
+				return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.Unauthorized, "authentication failed")
+			}
+			return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.ServerError, "server error while authenticating")
+		}
+	}
+
+	logrus.Debug("Binding username password")
+	searchRequest := ldapv2.NewSearchRequest(config.UserSearchBase,
+		ldapv2.ScopeWholeSubtree, ldapv2.NeverDerefAliases, 0, 0, false,
+		"(&("+ObjectClassAttribute+"="+config.UserObjectClass+")(uid="+username+"))", ldap.GetUserSearchAttributesForOpenLDAP(config), nil)
+	result, err := lConn.Search(searchRequest)
+	if err != nil {
+		return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.Unauthorized, "authentication failed") // need to reload this error
+	}
+
+	if len(result.Entries) < 1 {
+		return v3.Principal{}, nil, nil, fmt.Errorf("Cannot locate user information for %s", searchRequest.Filter)
+	} else if len(result.Entries) > 1 {
+		return v3.Principal{}, nil, nil, fmt.Errorf("ldap user search found more than one result")
+	}
+
+	userDN := result.Entries[0].DN //userdn is externalID
+	err = lConn.Bind(userDN, password)
+	if err != nil {
+		if ldapv2.IsErrorWithCode(err, ldapv2.LDAPResultInvalidCredentials) {
+			return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.Unauthorized, "authentication failed")
+		}
+		return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.ServerError, "server error while authenticating")
+	}
+
+	operationalAttrList := []string{"1.1", "+", "*"}
```

Comment:
what does this list do and why do we need it?
If needed, it should be move out of the function and stored as a global var since it it is always the same

Reply:
Yes we should store it as constant, this list is needed to fetch the operational attributes for the user object. This is needed for fetching group membership info.